06x-new-kafka-roller.md

```diff
@@ -0,0 +1,350 @@
+| maxRetries             | 10            | No              | The maximum number of times a node can be retried after not meeting the safety conditions. This is checked against the node's `numRetries`.                                                                                                                      |
+| postOperationTimeoutMs | 60 seconds    | Yes             | The maximum amount of time we will wait for nodes to transition to `SERVING` state after an operation in each retry. This will be based on the operation timeout that is already exposed to the user via environment variable `STRIMZI_OPERATION_TIMEOUT_MS`. |
+| maxRestartParallelism  | 1             | Yes             | The maximum number of broker nodes that can be restarted in parallel.                                                                                                                                                                                         |
+
+### Algorithm
+
+1. Initialise a context object for each node with the following data:
+```
+Context: {
+         nodeRef: <NodeRef object passed from KafkaReconciler>
+         nodeRoles: <This will be set using the pod labels `strimzi.io/controller-role` and `strimzi.io/broker-role>
+         state: UNKNOWN
+         lastTransition: <SYSTEM_TIME>
+         reason: <Result of the predicate function passed from KafkaReconciler>
+         numRestartAttempts: 0
+         numReconfigAttempts: 0
+         numRetries: 0
+}
+```
+Contexts are recreated in each reconciliation with the initial data above.
+
+2. Observe and transition each node's state to the corresponding state based on the information collected from the abstracted sources. If it can't retrieve any information from the sources, the reconciliation fails and the next reconciliation would start from step 1.
+
+3. If there are nodes in `NOT_READY` state, wait for them to have `SERVING` within the `postOperationTimeoutMs`.  
+   We want to give nodes chance to get ready before we try to connect to them or consider them for rolling. This is important especially for nodes which were just started.
+   This is consistent with how the current roller handles unready nodes.
+   - If the timeout is reached, proceed to the next step and check if any of the nodes need to be restarted.
+
+4. Group the nodes into the following categories based on their state and connectivity:
+   - `RESTART_FIRST` - Nodes that have `NOT_READY` or `NOT_RUNNING` state in their contexts. The nodes that we cannot connect to via Admin API will also be put into this group with its reason updated with `POD_UNRESPONSIVE`.
+   - `WAIT_FOR_LOG_RECOVERY` -  Nodes that have `RECOVERING` state.
+   - `RESTART` - Nodes that have non-empty list of reasons from the predicate function and have not been restarted yet (Context.numRestartAttempts == 0).
```

Comment:
Maybe `TO_BE_RESTARTED` sounds better?

Reply:
I'm not sure, then all of them would need to be changed with `TO_BE` prefix.

Reply:
+1 Tina

Reply:
Ok, but how we can make a better differentiation between `RESTART` and `RESTART_FIRST`. In general I think a state name should try to be clear, without the need to read its description in details.

Reply:
These are not state names, more like action names that nodes being categorised into but yes, we can certainly make a better differentiation between the 2. Perhaps, it could be `RESTART_UNREADY` and `RESTART_READY`?